server/events/apply_command_runner.go

```diff
@@ -0,0 +1,133 @@
+		disableApplyAll:     cmdRunner.DisableApplyAll,
+		commitStatusUpdater: cmdRunner.CommitStatusUpdater,
+		prjCmdBuilder:       cmdRunner.ProjectCommandBuilder,
+		prjCmdRunner:        cmdRunner.ProjectCommandRunner,
+	}
+}
+
+type ApplyCommandRunner struct {
+	cmdRunner           *DefaultCommandRunner
+	disableApplyAll     bool
+	vcsClient           vcs.Client
+	commitStatusUpdater CommitStatusUpdater
+	prjCmdBuilder       ProjectApplyCommandBuilder
+	prjCmdRunner        ProjectApplyCommandRunner
+}
+
+func (a *ApplyCommandRunner) Run(ctx *CommandContext, cmd *CommentCommand) {
+	var err error
+	baseRepo := ctx.Pull.BaseRepo
+	pull := ctx.Pull
+
+	if a.disableApplyAll && !cmd.IsForSpecificProject() {
+		ctx.Log.Info("ignoring apply command without flags since apply all is disabled")
+		if err := a.vcsClient.CreateComment(baseRepo, pull.Num, applyAllDisabledComment, models.ApplyCommand.String()); err != nil {
+			ctx.Log.Err("unable to comment on pull request: %s", err)
+		}
+
+		return
+	}
+
+	// Get the mergeable status before we set any build statuses of our own.
+	// We do this here because when we set a "Pending" status, if users have
+	// required the Atlantis status checks to pass, then we've now changed
+	// the mergeability status of the pull request.
+	ctx.PullMergeable, err = a.cmdRunner.VCSClient.PullIsMergeable(baseRepo, pull)
+	if err != nil {
+		// On error we continue the request with mergeable assumed false.
+		// We want to continue because not all apply's will need this status,
+		// only if they rely on the mergeability requirement.
+		ctx.PullMergeable = false
+		ctx.Log.Warn("unable to get mergeable status: %s. Continuing with mergeable assumed false", err)
+	}
+	ctx.Log.Info("pull request mergeable status: %t", ctx.PullMergeable)
+
+	if err = a.commitStatusUpdater.UpdateCombined(baseRepo, pull, models.PendingCommitStatus, cmd.CommandName()); err != nil {
+		ctx.Log.Warn("unable to update commit status: %s", err)
+	}
+
+	var projectCmds []models.ProjectCommandContext
+	projectCmds, err = a.prjCmdBuilder.BuildApplyCommands(ctx, cmd)
+
+	if err != nil {
+		if statusErr := a.commitStatusUpdater.UpdateCombined(ctx.Pull.BaseRepo, ctx.Pull, models.FailedCommitStatus, cmd.CommandName()); statusErr != nil {
+			ctx.Log.Warn("unable to update commit status: %s", statusErr)
+		}
+		a.cmdRunner.updatePull(ctx, cmd, CommandResult{Error: err})
+		return
+	}
+
+	// Only run commands in parallel if enabled
+	var result CommandResult
+	if a.isParallelEnabled(projectCmds) {
+		ctx.Log.Info("Running applies in parallel")
+		result = runProjectCmdsParallel(projectCmds, a.prjCmdRunner.Apply)
+	} else {
+		result = runProjectCmds(projectCmds, a.prjCmdRunner.Apply)
+	}
+
+	a.cmdRunner.updatePull(
+		ctx,
+		cmd,
+		result)
+
+	pullStatus, err := a.cmdRunner.updateDB(ctx, pull, result.ProjectResults)
+	if err != nil {
+		a.cmdRunner.Logger.Err("writing results: %s", err)
+		return
+	}
+
+	a.updateCommitStatus(ctx, pullStatus)
```

Comment:
This all seems like something we could package up in a nice generic implementation we can re-use across commands. Not saying we should do it as part of this PR, but just something to keep in mind.

Reply:
Yes, absolutely. I want to remove cyclic dependency from `DefaultCommandRunner`. But decided to do that later since this already was a bit of work. 
Candidates are `updatePull`, `updateCommitStatus`, `updateDB`, `automerge`, and `automergeEnabled`. At that point we should be able to move this refactor into a separate package.